exonum-java-binding/tutorials/car-registry/car-registry-service/src/main/java/com/example/car/MyService.java

```diff
@@ -0,0 +1,148 @@
+/* 
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.car;
+
+import com.example.car.messages.Transactions;
+import com.example.car.messages.Transactions.AddVehicle;
+import com.example.car.messages.VehicleOuterClass.Vehicle;
+import com.exonum.binding.core.blockchain.BlockchainData;
+import com.exonum.binding.core.runtime.ServiceInstanceSpec;
+import com.exonum.binding.core.service.AbstractService;
+import com.exonum.binding.core.service.Configuration;
+import com.exonum.binding.core.service.Node;
+import com.exonum.binding.core.storage.database.Prefixed;
+import com.exonum.binding.core.storage.indices.ProofMapIndexProxy;
+import com.exonum.binding.core.transaction.ExecutionException;
+import com.exonum.binding.core.transaction.Transaction;
+import com.exonum.binding.core.transaction.TransactionContext;
+import com.google.inject.Inject;
+import io.vertx.ext.web.Router;
+import java.util.List;
+import java.util.Optional;
+
+public final class MyService extends AbstractService {
+
+  public static final int ADD_VEHICLE_TX_ID = 0;
+  public static final int CHANGE_OWNER_TX_ID = 1;
+  public static final byte ID_ALREADY_EXISTS_ERROR_CODE = 100;
+  public static final byte NO_VEHICLE_ERROR_CODE = 101;
+
+  @Inject
+  public MyService(ServiceInstanceSpec instanceSpec) {
+    super(instanceSpec);
+  }
+
+  // ci-block ci-createPublicApiHandlers {
+  @Override
+  public void createPublicApiHandlers(Node node, Router router) {
+    var apiController = new ApiController(this, node);
+    apiController.mount(router);
+  }
+  // }
+
+  // ci-block ci-add-vehicle {
+  @Transaction(ADD_VEHICLE_TX_ID)
+  public void addVehicle(Transactions.AddVehicle args, TransactionContext context) {
+    var serviceData = context.getServiceData();
+    addVehicle(args, serviceData);
+  }
+
+  private void addVehicle(AddVehicle args, Prefixed serviceData) {
+    var schema = new MySchema(serviceData);
+    ProofMapIndexProxy<String, Vehicle> vehicles = schema.vehicles();
+
+    // Check there is no vehicle with such id in the registry yet
+    var newVehicle = args.getNewVehicle();
+    var id = newVehicle.getId();
+    if (vehicles.containsKey(id)) {
```

Comment:
`ExecutionPreconditions`?

Reply:
I decided against their use them for clarity (you see that an ExecutionException is being thrown).

Reply:
ExecutionException already references the preconditions at class-level Javadoc, we may add 

```java
 * @see ExecutionPreconditions
```

to each constructor as well.